### DISCUSSION TITLE
Is LocalSet#run_until cancel safe?

### DISCUSSION THREAD

**Discussion by bryanlarsen** [tokio-rs/tokio#6121] [Q&A] [answered]
I've got code that simplifies to something like this:

```
select! {
  local_set.run_until(async {
    mpsc_receiver.recv().await.unwrap()
  }) => {...},
  ...
}
```

And one of the local set tasks is putting messages onto the mpsc channel.

it's behaving oddly, the unwrap is panic'ing on a None when I wouldn't expect it to.

Most likely problem is PBKAC, but if LocalSet#run_until isn't cancel safe I'll need to do something else.

Can somebody verify that LocalSet#run_until is cancel-safe?   Either way this seems like something that should be in the docs.

Thanks.

**Comment by Darksonn** [comment, marked as answer]
Yes, as long as you give it a cancel safe future as the argument, `run_until` is cancel safe. Your issue is probably something else.

Please file a documentation bug about `run_until` not explaining this.

**Reply by bryanlarsen** [reply]
I simplified my code to `local_set.run_until(mspc_receiver.recv())` and I'm still getting the unexpected None.  mpsc::Receiver#recv is documented as cancel-safe, so problem must be elsewhere.   Thank you.

**Reply by bryanlarsen** [reply]
Found the issue in my code.
